pulp_ansible/app/sigstore/tasks/sigstore_signature.py

```diff
@@ -0,0 +1,295 @@
+
+
+def _generate_checksum_manifest(collection_version):
+    """
+    Generate a file containing the sha256 hashes of all the collection elements.
+    """
+    cartifact = collection_version.contentartifact_set.select_related("artifact").first()
+    artifact_name = cartifact.artifact.file.name
+    artifact_file = storage.open(artifact_name)
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        differ = DistlibManifestChecksumFileExistenceDiffer
+        checksum = ChecksumFile(tempdir, differ=differ)
+        with tarfile.open(fileobj=artifact_file, mode="r") as tar:
+            tar.extractall(path=tempdir)
+
+            try:
+                manifest = checksum.generate_gnu_style()
+
+            except FileNotFoundError as e:
+                if os.path.islink(e.filename):
+                    log.error(
+                        f"Broken symlink found at {e.filename} -- this is not supported. Aborting."
+                    )
+                if e.filename.endswith("/MANIFEST.in"):
+                    log.error("Could not find a MANIFEST.in file in the specified project.")
+                    log.info("If you are attempting to sign a project, please create this file.")
+                    log.info("See the ansible-sign documentation for more information.")
+                raise e
+
+            except DistlibException as e:
+                log.error(f"An error was encountered while parsing MANIFEST.in: {e}")
+                raise e
+
+            for warning in checksum.warnings:
+                log.warn(warning)
+            log.debug(
+                "Full calculated checksum manifest (%s):\n%s",
+                manifest,
+            )
+            return manifest
+
+
+def verify_sigstore_signature_upload(data):
+    # Check that the provided signature materials correspond to the collection content on upload.
+    """The task code for verifying Sigstore signature upload."""
+    collection = data["signed_collection"]
+    repository = data.get("repository")
+    if repository:
+        sigstore_verifying_service = repository.sigstore_verifying_service
+    else:
+        raise ValueError("This content type must be associated with a repository.")
+    signature = data["data"]
+    certificate = data["sigstore_x509_certificate"]
+
+    def format(pubkey):
+        delimiter = "-----"
+        s = pubkey.split(delimiter)
+        res = delimiter + s[1] + delimiter + re.sub(" +", "\n", s[2]) + delimiter + s[3] + delimiter
+        return res
+
+    certificate = format(certificate)
+    sigstore_bundle = data.get("sigstore_bundle")
+    checksums = _generate_checksum_manifest(collection)
+
+    bundle = Bundle().from_json(sigstore_bundle) if sigstore_bundle != "null" else None
+
+    with tempfile.NamedTemporaryFile(dir=".", delete=False, mode="w") as manifest_file:
+        manifest_file.write(checksums)
+        manifest_file.flush()
+    with open(manifest_file.name, "rb", buffering=0) as manifest_io:
```

Comment:
Since checksums is already in-memory no need to re-write it to a file, use io.BytesIO instead: https://docs.python.org/3/library/io.html#binary-i-o

Reply:
Fixed in https://github.com/pulp/pulp_ansible/pull/1428/commits/df18426680d102d00a3859e749fa9b075d422f99